flake: TestWebhookReinvocationPolicy

**Which jobs are failing**:

**Which test(s) are failing**:

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1172262471734398978

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1172263738850414593

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1172263738850414593

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1171840930756431874

```
=== RUN   TestWebhookReinvocationPolicyWithoutWatchCache/invalid_priority_class_set_by_webhook_should_result_in_error_from_in-tree_priority_plugin
    --- FAIL: TestWebhookReinvocationPolicyWithoutWatchCache/invalid_priority_class_set_by_webhook_should_result_in_error_from_in-tree_priority_plugin (0.02s)
        reinvocation_test.go:399: expected error but got none
```

```
=== RUN   TestWebhookReinvocationPolicyWithWatchCache/no_reinvocation_of_webhook_B_when_in-tree_or_prior_webhook_mutations#01
    --- FAIL: TestWebhookReinvocationPolicyWithWatchCache/no_reinvocation_of_webhook_B_when_in-tree_or_prior_webhook_mutations#01 (0.04s)
        reinvocation_test.go:416: expected labels 'map[a:true x:true]', but got 'map[fight:false x:true]'
        reinvocation_test.go:423: expected 1 invocations of /addlabel, but got 0
        reinvocation_test.go:438: failed to get expected events -- missing: missing 1 events
            
            - first event checked: &audit.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, Level:"Request", AuditID:"d435699a-23c2-4831-8f30-a0d1691e55cf", Stage:"RequestReceived", RequestURI:"/api/v1/namespaces/default/pods/marker", Verb:"patch", User:v1.UserInfo{Username:"system:apiserver", UID:"6e4aace2-3ec9-481b-b475-7439a0d22ad3", Groups:[]string{"system:masters"}, Extra:map[string]v1.ExtraValue(nil)}, ImpersonatedUser:(*v1.UserInfo)(nil), SourceIPs:[]string{"127.0.0.1"}, UserAgent:"admissionwebhook.test/v0.0.0 (linux/amd64) kubernetes/$Format", ObjectRef:(*audit.ObjectReference)(0xc03d3b1900), ResponseStatus:(*v1.Status)(nil), RequestObject:(*runtime.Unknown)(nil), ResponseObject:(*runtime.Unknown)(nil), RequestReceivedTimestamp:v1.MicroTime{Time:time.Time{wall:0x29974ef0, ext:63703922159, loc:(*time.Location)(0xaf69640)}}, StageTimestamp:v1.MicroTime{Time:time.Time{wall:0x29974ef0, ext:63703922159, loc:(*time.Location)(0xaf69640)}}, Annotations:map[string]string(nil)}
            
            - last event checked: &audit.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, Level:"Request", AuditID:"f061a270-57bc-4356-a34f-370e46d9f0b7", Stage:"ResponseComplete", RequestURI:"/api/v1/namespaces/reinvoke-5/pods", Verb:"create", User:v1.UserInfo{Username:"system:apiserver", UID:"6e4aace2-3ec9-481b-b475-7439a0d22ad3", Groups:[]string{"system:masters"}, Extra:map[string]v1.ExtraValue(nil)}, ImpersonatedUser:(*v1.UserInfo)(0xc032a330c0), SourceIPs:[]string{"127.0.0.1"}, UserAgent:"admissionwebhook.test/v0.0.0 (linux/amd64) kubernetes/$Format", ObjectRef:(*audit.ObjectReference)(0xc03d3b1a80), ResponseStatus:(*v1.Status)(0xc038b14280), RequestObject:(*runtime.Unknown)(0xc040285860), ResponseObject:(*runtime.Unknown)(nil), RequestReceivedTimestamp:v1.MicroTime{Time:time.Time{wall:0x29f47e68, ext:63703922159, loc:(*time.Location)(0xaf69640)}}, StageTimestamp:v1.MicroTime{Time:time.Time{wall:0x2b35c480, ext:63703922159, loc:(*time.Location)(0xaf69640)}}, Annotations:map[string]string{"authorization.k8s.io/decision":"allow", "authorization.k8s.io/reason":"", "mutation.webhook.admission.k8s.io/round_0_index_0":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.0.settrue\",\"mutated\":true}", "mutation.webhook.admission.k8s.io/round_0_index_1":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.1.setfalse\",\"mutated\":true}", "mutation.webhook.admission.k8s.io/round_1_index_0":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.0.settrue\",\"mutated\":true}", "mutation.webhook.admission.k8s.io/round_1_index_1":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.1.setfalse\",\"mutated\":true}", "patch.webhook.admission.k8s.io/round_0_index_0":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.0.settrue\",\"patch\":[{\"op\":\"replace\",\"path\":\"/metadata/labels/fight\",\"value\":\"true\"}],\"patchType\":\"JSONPatch\"}", "patch.webhook.admission.k8s.io/round_0_index_1":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.1.setfalse\",\"patch\":[{\"op\":\"replace\",\"path\":\"/metadata/labels/fight\",\"value\":\"false\"}],\"patchType\":\"JSONPatch\"}", "patch.webhook.admission.k8s.io/round_1_index_0":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.0.settrue\",\"patch\":[{\"op\":\"replace\",\"path\":\"/metadata/labels/fight\",\"value\":\"true\"}],\"patchType\":\"JSONPatch\"}", "patch.webhook.admission.k8s.io/round_1_index_1":"{\"configuration\":\"admission.integration.test-4\",\"webhook\":\"admission.integration.test.1.setfalse\",\"patch\":[{\"op\":\"replace\",\"path\":\"/metadata/labels/fight\",\"value\":\"false\"}],\"patchType\":\"JSONPatch\"}"}}
            
            - number of events checked: 4
            
            - missing events: []utils.AuditEvent{utils.AuditEvent{ID:"", Level:"Request", Stage:"ResponseComplete", RequestURI:"/api/v1/namespaces/reinvoke-5/pods", Verb:"create", Code:201, User:"system:apiserver", ImpersonatedUser:"webhook-reinvocation-integration-client", ImpersonatedGroups:"system:authenticated,system:masters", Resource:"pods", Namespace:"reinvoke-5", RequestObject:true, ResponseObject:false, AuthorizeDecision:"allow", AdmissionWebhookMutationAnnotations:map[string]string{"mutation.webhook.admission.k8s.io/round_0_index_1":"{\"configuration\":\"admission.integration.test-5\",\"webhook\":\"admission.integration.test.1.addlabel\",\"mutated\":true}"}, AdmissionWebhookPatchAnnotations:map[string]string{"patch.webhook.admission.k8s.io/round_0_index_1":"{\"configuration\":\"admission.integration.test-5\",\"webhook\":\"admission.integration.test.1.addlabel\",\"patch\":[{\"op\":\"add\",\"path\":\"/metadata/labels/a\",\"value\":\"true\"}],\"patchType\":\"JSONPatch\"}"}}}
```

**Since when has it been failing**:

**Testgrid link**:

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-integration&sort-by-flakiness=&include-filter-by-regex=WebhookReinvocation&width=5

**Reason for failure**:

**Anything else we need to know**:

/sig api-machinery
/assign @jpbetz

--- a/test/integration/apiserver/admissionwebhook/reinvocation_test.go
+++ b/test/integration/apiserver/admissionwebhook/reinvocation_test.go
@@ -26,6 +26,7 @@ import (
 	"net/http/httptest"
 	"os"
 	"reflect"
+	"strconv"
 	"strings"
 	"sync"
 	"testing"
@@ -312,25 +313,36 @@ func testWebhookReinvocationPolicy(t *testing.T, watchCache bool) {
 		}
 	}
 
-	_, err = client.CoreV1().Pods("default").Create(reinvocationMarkerFixture)
-	if err != nil {
-		t.Fatal(err)
-	}
-
 	for i, tt := range testCases {
 		t.Run(tt.name, func(t *testing.T) {
 			upCh := recorder.Reset()
-			ns := fmt.Sprintf("reinvoke-%d", i)
-			_, err = client.CoreV1().Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ns}})
+			testCaseID := strconv.Itoa(i)
+			ns := "reinvoke-" + testCaseID
+			nsLabels := map[string]string{"test-case": testCaseID}
+			_, err = client.CoreV1().Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ns, Labels: nsLabels}})
 			if err != nil {
 				t.Fatal(err)
 			}
 
+			// Write markers to a separate namespace to avoid cross-talk
+			markerNs := ns + "-markers"
+			markerNsLabels := map[string]string{"test-markers": testCaseID}
+			_, err = client.CoreV1().Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: markerNs, Labels: markerNsLabels}})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Create a maker object to use to check for the webhook configurations to be ready.
+			marker, err := client.CoreV1().Pods(markerNs).Create(newReinvocationMarkerFixture(markerNs))
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			fail := admissionv1beta1.Fail
 			webhooks := []admissionv1beta1.MutatingWebhook{}
 			for j, webhook := range tt.webhooks {
-				name := fmt.Sprintf("admission.integration.test.%d.%s", j, strings.TrimPrefix(webhook.path, "/"))
-				fail := admissionv1beta1.Fail
 				endpoint := webhookServer.URL + webhook.path
+				name := fmt.Sprintf("admission.integration.test.%d.%s", j, strings.TrimPrefix(webhook.path, "/"))
 				webhooks = append(webhooks, admissionv1beta1.MutatingWebhook{
 					Name: name,
 					ClientConfig: admissionv1beta1.WebhookClientConfig{
@@ -342,11 +354,28 @@ func testWebhookReinvocationPolicy(t *testing.T, watchCache bool) {
 						Rule:       admissionv1beta1.Rule{APIGroups: []string{""}, APIVersions: []string{"v1"}, Resources: []string{"pods"}},
 					}},
 					ObjectSelector:          webhook.objectSelector,
+					NamespaceSelector:       &metav1.LabelSelector{MatchLabels: nsLabels},
 					FailurePolicy:           &fail,
 					ReinvocationPolicy:      webhook.policy,
 					AdmissionReviewVersions: []string{"v1beta1"},
 				})
 			}
+			// Register a marker checking webhook with each set of webhook configurations
+			markerEndpoint := webhookServer.URL + "/marker"
+			webhooks = append(webhooks, admissionv1beta1.MutatingWebhook{
+				Name: "admission.integration.test.marker",
+				ClientConfig: admissionv1beta1.WebhookClientConfig{
+					URL:      &markerEndpoint,
+					CABundle: localhostCert,
+				},
+				Rules: []admissionv1beta1.RuleWithOperations{{
+					Operations: []admissionv1beta1.OperationType{admissionv1beta1.OperationAll},
+					Rule:       admissionv1beta1.Rule{APIGroups: []string{""}, APIVersions: []string{"v1"}, Resources: []string{"pods"}},
+				}},
+				NamespaceSelector:       &metav1.LabelSelector{MatchLabels: markerNsLabels},
+				ObjectSelector:          &metav1.LabelSelector{MatchLabels: map[string]string{"marker": "true"}},
+				AdmissionReviewVersions: []string{"v1beta1"},
+			})
 
 			cfg, err := client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Create(&admissionv1beta1.MutatingWebhookConfiguration{
 				ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("admission.integration.test-%d", i)},
@@ -364,7 +393,7 @@ func testWebhookReinvocationPolicy(t *testing.T, watchCache bool) {
 
 			// wait until new webhook is called the first time
 			if err := wait.PollImmediate(time.Millisecond*5, wait.ForeverTestTimeout, func() (bool, error) {
-				_, err = client.CoreV1().Pods("default").Patch(reinvocationMarkerFixture.Name, types.JSONPatchType, []byte("[]"))
+				_, err = client.CoreV1().Pods(markerNs).Patch(marker.Name, types.JSONPatchType, []byte("[]"))
 				select {
 				case <-upCh:
 					return true, nil
@@ -529,17 +558,15 @@ func newReinvokeWebhookHandler(recorder *invocationRecorder) http.Handler {
 			http.Error(w, err.Error(), 400)
 		}
 
-		// When resetting between tests, a marker object is patched until this webhook
-		// observes it, at which point it is considered ready.
-		if pod.Namespace == reinvocationMarkerFixture.Namespace && pod.Name == reinvocationMarkerFixture.Name {
-			recorder.MarkerReceived()
-			allow(w)
-			return
-		}
-
 		recorder.IncrementCount(r.URL.Path)
 
 		switch r.URL.Path {
+		case "/marker":
+			// When resetting between tests, a marker object is patched until this webhook
+			// observes it, at which point it is considered ready.
+			recorder.MarkerReceived()
+			allow(w)
+			return
 		case "/noop":
 			allow(w)
 		case "/settrue":
@@ -600,15 +627,20 @@ func expectedAuditEvents(webhookMutationAnnotations, webhookPatchAnnotations map
 	}
 }
 
-var reinvocationMarkerFixture = &corev1.Pod{
-	ObjectMeta: metav1.ObjectMeta{
-		Namespace: "default",
-		Name:      "marker",
-	},
-	Spec: corev1.PodSpec{
-		Containers: []v1.Container{{
-			Name:  "fake-name",
-			Image: "fakeimage",
-		}},
-	},
+func newReinvocationMarkerFixture(namespace string) *corev1.Pod {
+	return &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "marker",
+			Labels: map[string]string{
+				"marker": "true",
+			},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []v1.Container{{
+				Name:  "fake-name",
+				Image: "fakeimage",
+			}},
+		},
+	}
 }

